### PR TITLE
api: new field to the FilterConfig to disable a filter in specific route or vh

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: api: new field to the FilterConfig to disable a filter in specific route or vh
Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
draft idea to the #20867

**Comment by htuch** [comment]
Is this just a generic filter disable at round config override level? Seems reasonable to me. CC @envoyproxy/api-shepherds

**Comment by wbpcode** [comment]
> Is this just a generic filter disable at round config override level? Seems reasonable to me. CC @envoyproxy/api-shepherds

Yeap. By this way, we can disable any filter in a specific route and needn't change any API of the filter. It means that the different routes could execute complete different filter chains.

**Comment by adisuissa** [comment]
Thanks, this LGTM, and I like the idea of making it general across all filters.
One thing to note is what should happen to "Overridden" configs (some also have their own "disabled" flag) that have this flag set to true.
IMHO the config should be rejected, but not sure how to implement this in a better way other than looking at all filters adding this validation.

> ideally, we can skip the disabled filters' creation when we create the filter chain. Then the disable filters will not bring any additional overhead at runtime.
That would be nice, but I'm not sure how this can be done for a per-route override. Some other route may need that filter, and so it is challenging to know which filters to instantiate a-priori.

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -2384,4 +2384,8 @@ message FilterConfig {
   // not support the specified filter, it may ignore the map entry rather
   // than rejecting the config.
   bool is_optional = 2;
+
+  // If true, the filter is disabled in the route or virtual host and the
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply api/envoy/config/route/v3/route_components.proto:2388]: Wow, you got it.  that's what I want completely. 👍 

Although, in my original opinion, the behavior depend on the implementation so I initially intend to document it more detailedly in a seperated doc.

**Comment by kyessenov** [comment]
This sounds reasonable to me too, as an alternative to composite filters for the typed_per_filter_configs, as a way to exempt some traffic from filters.

**Comment by wbpcode** [comment]
More detailed comment is added.

Advantages of this way:
- Better performance. We can evaluate all these disable flag at config loading phase and skip unnecessary filter creations. And the disabled filters will skipped completely, no any `decode*`/`encode*` methods will be called.
- Less memory overhead at runtime. Disabled filters will not be created.
- Generic way to disable any filter and no any code/API change is necessary to the target filter.
- Route level filter chain. Customize filter chain by enable/disable different filters set. (May be we can add a flag to tell envoy only add the explictly enabled filters to the filter chain in the future. 🤔 It will be helpful if there are lots of filters and most of routes only need one or two of these filters, for example enable cors for route A, enable ext_authz for route B, and enable cache for route C, etc.)

Disadvantages of this way:
- Not friendly to the route refresh.
- Some more memory to store the enable/disable flag.

**Comment by kyessenov** [comment]
@wbpcode I don't recall - does this require that a filter explicitly implements "most specific filter config" in code, or is this automatic?

**Comment by wbpcode** [comment]
> @wbpcode I don't recall - does this require that a filter explicitly implements "most specific filter config" in code, or is this automatic?

This is automatic and it affects the filter chain directly, acting outside the filter.

File: api/envoy/config/route/v3/route_components.proto [hunk 2]
```diff
@@ -2384,4 +2384,18 @@ message FilterConfig {
...
+  //
+  //   This field will take effect when the request arrive and filter chain is created for the request.
+  //   If initial route is selected for the request and a filter is disabled in the initial route, then
+  //   the filter will not be added to the filter chain.
+  //   And if the request is mutated later and re-match to another route, the disabled filter by the
+  //   initial route will not be added back to the filter chain because the filter chain is already
+  //   created and it is too late to change the chain.
+  //
+  //   This field only make sense for the downstream HTTP filters for now.
+  bool is_disabled = 3;
```
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:2400]: Nit: I would just call this `disabled`. I think `is_optional` above is a workaround for the fact that `optional` is a reserved proto keyword.

**Review by htuch — COMMENTED** [review]
LGTM

File: api/envoy/config/route/v3/route_components.proto (hunk 2 above)
Comment (wbpcode) [reply api/envoy/config/route/v3/route_components.proto]: done.

**Comment by wbpcode** [comment]
Seems there is no explict objection to the API and functions. I think we can merge it and then I can prepare a prototype of the implementation in the next days. 😄

**Comment by wbpcode** [comment]
A stamp from maintainer is still necessary. Friendly ping @htuch 😺

File: api/envoy/config/route/v3/route_components.proto [hunk 3]
```diff
@@ -2384,4 +2384,18 @@ message FilterConfig {
...
+  // .. note::
+  //
+  //   This field will take effect when the request arrive and filter chain is created for the request.
+  //   If initial route is selected for the request and a filter is disabled in the initial route, then
+  //   the filter will not be added to the filter chain.
+  //   And if the request is mutated later and re-match to another route, the disabled filter by the
+  //   initial route will not be added back to the filter chain because the filter chain is already
+  //   created and it is too late to change the chain.
+  //
+  //   This field only make sense for the downstream HTTP filters for now.
```
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:2399]: Should add a not-implemented-yet annotation or merge this into the implementation PR.

Comment (wbpcode) [reply api/envoy/config/route/v3/route_components.proto:2399]: The `FilterConfig` message are annotated as `not-implemented`. So, it's unnecessary to add the annotation seperatedly for this field.

**Review by htuch — APPROVED** [review]
/lgtm api

**Pull request merged**